### PR TITLE
feat: add a feature to enable jemalloc profiling

### PR DESCRIPTION
**Pull request by yangby-cryptape**
- [x] fix: fix a ci script which check redundant dependencies
- [x] feat: run a memory tracker thread to record the memory statistics
  - A thread will be started in background to monitor the whole process, and trace the memory.
- [x] feat: add a feature to enable jemalloc profiling
  - [CKB Debugging](https://github.com/nervosnetwork/ckb/blob/cb2bbd9073e1d6bbddc20eb5ab900e154539ab9d/docs/ckb-debugging.md) (Document about this feature.)

### CODE REVIEW

**Comment by yangby-cryptape** [comment]
I rebase this PR and push the last commit.

File: Cargo.toml [hunk 1]
```diff
@@ -69,9 +69,12 @@ members = [
 [profile.release]
 overflow-checks = true
 
-[target.'cfg(all(not(target_env = "msvc"), not(target_os="macos")))'.dependencies]
+[target.'cfg(all(not(target_env = "msvc"), not(target_os="macos"), not(feature = "profiling")))'.dependencies]
 jemallocator = { version = "0.3.0" }
+[target.'cfg(all(not(target_env = "msvc"), not(target_os="macos"), feature = "profiling"))'.dependencies]
```
Comment (xxuejie) [inline Cargo.toml]: The doc below mentioned that this feature is only available on Linux, so why not test target_os is Linux directly? Does the config mean that toolchain that is not using msvc(e.g: mingw) will work on Windows?

Comment (yangby-cryptape) [reply Cargo.toml]: Beats me! I'm confused, too.

I just keep the same with `Cargo.toml`.
https://github.com/nervosnetwork/ckb/blob/dad394ea3f50f518a40e5a8a457dfb6811ba245a/Cargo.toml#L72-L73

This two lines are introduced by #1799. Maybe @jjyr could answer this.

File: Makefile [hunk 2]
```diff
@@ -77,6 +77,10 @@ check: setup-ckb-test ## Runs all of the compiler's checks.
 build: ## Build binary with release profile.
 	cargo build ${VERBOSE} --release
 
+.PHONY: build
```
Comment (driftluo) [inline Makefile]: may  typo, should be "build-for-profiling"

Comment (yangby-cryptape) [reply Makefile]: Yes, it's a typo. I will force-push this PR to fix it.

File: Cargo.toml (hunk 1 above)
Comment (jjyr) [reply Cargo.toml]: I think I just copied it from somewhere, can't find the first place.

After checking the `jemallocator` crate, I belieave we should use the `cfg(all(target_arch = "x86_64",target_os="linux"))` compile condition, which only enable `jemalloc` on the tier1 supports platform.

https://crates.io/crates/jemallocator

Comment (yangby-cryptape) [reply Cargo.toml]: - It's copied from [README.md of jemallocator](https://github.com/gnzlbg/jemallocator/blob/c27a859e98e3cb790dc269773d9da71a1e918458/README.md#documentation).

  - We copied `target_env = "msvc"`.
  - And we appended `target_os="macos"`.

- [And jemallocator can be compiled on Windows without msvc.](https://github.com/gnzlbg/jemallocator/blob/c27a859e98e3cb790dc269773d9da71a1e918458/README.md#platform-support) So the official `README.md` suggests `target_env = "msvc"`.

  | Windows targets                      | build     | run      | tests           | valgrind     |
  |-------------------------------------|-----------|---------|--------------|--------------|
  | `x86_64-pc-windows-msvc` (tier 1)   | ✗         | ✗       | ✗            | ✗            |
  | `i686-pc-windows-msvc` (tier 1)     | ✗         | ✗       | ✗            | ✗            |
  | `x86_64-pc-windows-gnu` (tier 1)    | ✓         | ✓       | ✗            | ✗            |
  | `i686-pc-windows-gnu` (tier 1)      | ✓         | ✓       | ✗            | ✗            |

File: ckb-bin/src/subcommand/run.rs [hunk 3]
```diff
@@ -142,7 +144,8 @@ pub fn run(args: RunArgs, version: Version) -> Result<(), ExitCode> {
         .enable_experiment(shared.clone())
         .enable_integration_test(shared.clone(), network_controller.clone(), chain_controller)
         .enable_alert(alert_verifier, alert_notifier, network_controller)
-        .enable_indexer(&args.config.indexer, shared.clone());
+        .enable_indexer(&args.config.indexer, shared.clone())
+        .enable_debug();
```
Comment (jjyr) [inline ckb-bin/src/subcommand/run.rs]: ??

**Comment by doitian** [comment]
bors r=doitian,jjyr,driftluo

**Comment by doitian** [comment]
Version unmatch:

```
$ make check-cargotoml
./devtools/ci/check-cargotoml.sh
[BEGIN] Checking Cargo.toml ...
Error: Version in <./util/memory-tracker/Cargo.toml> is not right (expect: '0.30.1-pre', actual: '0.28.1-pre')
Warn: [dependencies::phf] in <./resource>
[ END ] Found 1 errors.
Makefile:141: recipe for target 'check-cargotoml' failed
make: *** [check-cargotoml] Error 1
The command "make check-cargotoml" exited with 2.
```

**Comment by yangby-cryptape** [comment]
The last force-pushed was a mistake.
I checked the diff view of force-pushed, I found only one line changed, I thought I committed a wrong commit.
The fact is: the diff view of force-pushed only display the different between last commit and previous commit.

The last three force-pushed only updated two line:
- Version of `ckb-memory-tracker` in `Cargo.toml`.
- Version of `ckb-memory-tracker` in `Cargo.lock`.

**Comment by doitian** [comment]
heim-net has locked hex to a yanked version

<https://github.com/heim-rs/heim/blob/ba88f7de42f0fb579fe38de596716c009816c768/heim-net/Cargo.toml#L32>

**Pull request merged**